### ISSUE TITLE
Using sockets equal to client <-> server connection?

### ISSUE THREAD

**Issue by gatoniel** [containers/podman-py#15]
Hi,

I used docker containers for some time, but didn´t bother the underlying technicalities. I came accross Podman and liked some of its underlying differing techniques. I hope I won´t make a complete fool out of me for this issue...

As I understand it, one huge drawback of docker is, that it uses a client <-> server structure. From a security point, this implies that it is hard to follow who issued commands: A docker command from the commandline is sent to the server and the server creates a new container. Since the server creates the container, the system can´t audit who issued the former command on the commandline and hence we have a security problem. Podman in contrast does not have this client <-> server structure. The command issued on the commandline directly creates the container; additionaly we don´t need any daemon running in the background. This might be an oversimplified summary of what I read in articles about differences of docker and Podman.

If I understand the code here correct, this project will use an ApiConnector that communicates via sockets with another program. Latter program will create containers, when the python programm wants it to. But isn´t this again a server <-> client communication, where we need a program listening on the socket?

I do understand, that one needs a client <-> server communication, when one wants to communicate with a remote host. But when a python application wants to run a container on the same system, wouldn´t it be more "safe", to execute the podman command as a subprocess?

**Comment by rhatdan** [comment]
BTW, this type of question/discussion would be better at the podman mailing list podman.io

But anyways...

Yes traditional podman client is using fork/exec model, which gives you the advantages you were talking about.  For a lot of cases python directly execing the podman command would give you the same features.  But this is not an API, some users do not like screen scraping.  IE Trying to interpret output from the podman CLI can be difficult.  Therefore we need a remote API.  This Remote API allows us to interact with programs running as Python, or Javascript (cockpit) or even with remote go programs, which is being developed now.  

We also want Podman to replace Docker use cases that use docker-py.  Lots of infrastructure has been built to use this and we can not replace Docker fully until we support its API.  This also makes converting docker-compose programs a lot easier.

Two other features of the podman APIV2 is that it is socket activated, so their is no Daemon (other then systemd) and each connection gets its own Podman.  Because of this we can even support rootless podman.  So you can launch containers remotely without requiring a root running daemon.

If you want to continue the discussion lets do it on the podman.io mailing list.
